cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationLifecycleManager.java

```diff
@@ -72,19 +78,48 @@ public boolean scanOperations(ScanOperationRunsRequest request, int txBatchSize,
     return currentLimit == 0;
   }
 
+  /**
+   * Runs a given operation. It is the responsibility of the caller to validate state transition.
+   *
+   * @param detail {@link OperationRunDetail} of the operation
+   */
+  public OperationController startOperation(OperationRunDetail detail) {
+    return runtime.run(detail);
+  }
+
 
   /**
-   * Scan all pending operations. Needed for try running all pending operation during startup.
+   * Initiate operation stop. It is the responsibility of the caller to validate state transition.
    *
-   * @param consumer {@link Consumer} to process each scanned run
+   * @param detail {@link OperationRunDetail} of the operation
+   * @throws IllegalStateException in case the operation is not running. It is the
+   *     responsibility of the caller to handle the same
    */
-  public void scanPendingOperations(Consumer<OperationRunDetail> consumer)
-      throws IOException, InvalidFieldException {
-    TransactionRunners.run(transactionRunner, context -> {
-      getOperationRunStore(context).scanOperationByStatus(OperationRunStatus.PENDING, consumer);
-    }, IOException.class, InvalidFieldException.class);
+  public void stopOperation(OperationRunDetail detail) {
+    OperationController controller = runtime.getController(detail);
+    if (controller == null) {
+      throw new IllegalStateException("Operation is not running");
+    }
+    controller.stop();
   }
 
+  /**
+   * Resume the operation. Called after service restart.
+   *
+   * @param detail {@link OperationRunDetail} of the operation
+   */
+  public void resumeOperation(OperationRunDetail detail, OperationStatePublisher statePublisher) {
```

Comment:
couldnot find a better name. Any suggestion would be appreiciated.

Reply:
I think we can name it isRunning() and have it just check if the controller exists. Then have the caller publish the failed state if it's not running.